drivers: ethernet: stm32: Revert complete v1/v2 split and instanciate factorisation

erwango opened the pull request:
Revert PR "drivers: ethernet: eth_stm32_hal split v1 and v2" (#94192) which has been merged before a consensus was reached.
Instead provide an alternative split which introduces a `_common.c` file and hence maximizes code factorization to ease maintenance.
While I'm at, split ptp code in a dedicated file as well.

erwango committed:
Revert "drivers: ethernet: eth_stm32 v1/v2 are placed in separate files"

This reverts commit 29750ac05de52ffe0dc2873c5de4f261001f3b31.
This commit was part of PR 94142 which has been merged
while a discussion was still on going on the right direction
between splitting V1 and V2 or keeping a common file.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
Revert "drivers: ethernet: rename eth_stm32_hal.c -> eth_stm32_hal_v2.c"

This reverts commit 678e226045f64cfa084239ae746109a1613aa7ad.
This commit was part of PR 94142 which has been merged
while a discussion was still on going on the right direction
between splitting V1 and V2 or keeping a common file.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
drivers: ethernet: Rename eth_stm32_hal.c to eth_stm32hal_common.c

Keep the history on the common file before splitting v1 and v2 code from
it.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
drivers: ethernet: stm32: Split v1 and v2 code

Split V1 and V2 code from the common file of STM32 ethernet
driver.
Whenever it is possible, make use of v1 or v2 version of functions
that are call from the common file in order to maximize factorization.

Review definitions to put them where they are used.
Review header inclusions to minimize them.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
drivers: ethernet: stm32: Split ptp functions into a dedicated source file

In an effort to make stm32 eth driver files easier to read and maintain,
split the ptp code in a dedicated file.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
drivers: ethernet: stm32: Rename functions that are not static anymore

To be able to split code, some functions were removed their static
attribute, which opens the possibility of name conflict.
Rename them by adding eth_stm32_ prefix.

Additionally, review their sorting in the header file.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -45,84 +23,18 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
-							ETH_DMATXDESC_OWN)
-
-#endif /* DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_ethernet) */
+#define LOG_MODULE_NAME eth_stm32_hal_v2
```
shouldn't the stm32 eth driver share one module name?

erwango replied:
Done

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -1,34 +1,93 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+extern struct eth_stm32_hal_dev_data eth0_data;
```
I don't like this being shared over multiple files, should be set near `ETH_NET_DEVICE_DT_INST_DEFINE` in preparation for adding support for multiple instance support

erwango replied:
Done

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -0,0 +1,484 @@
+
+const struct device *eth_stm32_phy_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0, phy_handle));
+
+struct net_if *eth_stm32_get_iface(struct eth_stm32_hal_dev_data *ctx)
+{
+	return ctx->iface;
+}
```
having this as a non static function, seems to add overhead, why not remove it and access it directly

erwango replied:
Done

erwango commented on drivers/ethernet/eth_stm32_hal_common.c:
```diff
@@ -0,0 +1,484 @@
+
+	ethernet_init(iface);
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+	/* This function requires the Ethernet interface to be
+	 * properly initialized. In auto-negotiation mode, it reads the speed
+	 * and duplex settings to configure the driver accordingly.
+	 */
+	eth_stm32_hal_init(dev);
+#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
```
Good point, I wondered where the difference came from. Just need to test it.

erwango committed:
drivers: ethernet: stm32: Remove eth_stm32_get_iface

Now that the function is not static anymore, it just adds useless overhead.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango committed:
drivers: ethernet: Move back hal_init_v2 to driver init

Since phy init is not in this function anymore, it can be moved back
to driver initialization function.

While I'm at, harmonize v1/v2 flavors of the function.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v1.c:
```diff
@@ -0,0 +1,319 @@
+	k_sleep(K_MSEC(1));
+	heth->Instance->MACFFR = tmp;
+}
+
```
Could you remove this extra empty line?

erwango replied:
Done

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -0,0 +1,637 @@
+#include <soc.h>
+#include <zephyr/irq.h>
+
+#if defined(CONFIG_PTP_CLOCK_STM32_HAL)
```
Is the if-defined condition really needed?

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -0,0 +1,637 @@
+}
+
+/* Pointer to an array of ETH_STM32_RX_BUF_SIZE uint8_t's */
+typedef uint8_t (*RxBufferPtr)[ETH_STM32_RX_BUF_SIZE];
```
Could you move this definition before line 46?

erwango replied:
Done

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -0,0 +1,637 @@
+
+	header->size = Length;
+
+	if (!*pStart) {
```
Would prefer `	if (*pStart == NULL) {` Cleanup in a later change?

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -0,0 +1,637 @@
+}
+
+/* allocate a tx buffer and mark it as used */
+static inline uint16_t allocate_tx_buffer(void)
```
All `inline` keywords here are useless. In a later change?

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -0,0 +1,637 @@
+}
+
+#if defined(CONFIG_ETH_STM32_HAL_API_V2)
+static ETH_TxPacketConfig tx_config;
```
This is used only far below. Move to source file top, next to line 46?

erwango replied:
Done

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -0,0 +1,637 @@
+	}
+
+	res = 0;
+error:
```
Since this is not an error only path, it should better be renamed `out`. In a later change?

etienne-lms commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -141,20 +136,24 @@ struct eth_stm32_hal_dev_data {
 void eth_stm32_mcast_filter(const struct device *dev,
 			    const struct ethernet_filter *filter);
 #endif /* CONFIG_ETH_STM32_MULTICAST_FILTER */
+#ifdef CONFIG_PTP_CLOCK_STM32_HAL
```
an empty line above would be welcome IMHO. Ditto at line 150.

erwango replied:
Done

etienne-lms reviewed (commented):
LGTM with comments to address now or later...

erwango commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -497,7 +497,7 @@ int eth_stm32_hal_init(const struct device *dev)
 		/* a recoverable error. Log the issue and continue */
 		/* driver initialisation */
-		LOG_ERR("HAL_ETH_Init Timed out");
+		LOG_WRN("HAL_ETH_Init timed out (cable not connected?)");
```
Thanks, done

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -1,34 +1,104 @@
+
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
should only be registered exactly ones the other times it should only be declared

erwango replied:
So, I'm duplicating all LOG modules declaration to have one per c file, with each its own name. Is that fine for both of you ? @maass-hamburg, @etienne-lms

etienne-lms replied:
Okay for me:
*eth_stm32_hal_common.c*: `LOG_MODULE_REGISTER(eth_stm32_hal, CONFIG_ETHERNET_LOG_LEVEL);`
*eth_stm32_hal_v1.c*: `LOG_MODULE_DECLARE(eth_stm32_hal);`
*eth_stm32_hal_v2.c*: `LOG_MODULE_DECLARE(eth_stm32_hal);`
*eth_stm32_hal_ptp.c*: `LOG_MODULE_REGISTER(eth_stm32_hal_ptp, CONFIG_ETHERNET_LOG_LEVEL);`

erwango commented on drivers/ethernet/eth_stm32_hal_ptp.c:
```diff
@@ -0,0 +1,344 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/ethernet.h>
+#include <soc.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include <zephyr/drivers/ptp_clock.h>
+
+#define LOG_MODULE_NAME eth_stm32_hal_ptp
+
+#include "eth_stm32_hal_priv.h"
```
This won't work as `priv.h` calls `LOG_MODULE_REGISTER(LOG_MODULE_NAME);` which expects `LOG_MODULE_NAME` to be defined beforehand.
This being said, I can just remove `eth_stm32_hal_ptp` and have all files using same `eth_stm32_hal` module name, as previously

etienne-lms commented on drivers/ethernet/eth_stm32_hal_priv.h:
```diff
@@ -9,14 +9,20 @@
+	bool used;
+};
+
+#endif /* CONFIG_ETH_STM32_HAL_API_V2 */
```
OK

erwango commented on drivers/ethernet/eth_stm32_hal_ptp.c:
```diff
@@ -0,0 +1,344 @@
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include <zephyr/drivers/ptp_clock.h>
```
This builds. Verified in in last push as well.

maass-hamburg commented:
I generally would have preferred it, if the shared functions been put directly in a shared header and keeping the dev defines in the v1 and v2 files. When making the v2 instance based, we wouldn't need to do the same for v1 and would only need minor changes. Making it this way is not really making it more readable than before, they were split.

erwango committed:
drivers: ethernet: stm32: #include sections cleanup

Clean up headers includes section:
- Remove conditional inclusions
- zephyr/, then susbystem/, standardlib, locals
- Sort alphabetically within these categories
- Remove the ones already available through _priv.h

Noticeable exception, lldp.h requires prior net_pkt.h inclusion

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

etienne-lms reviewed (approved):
LGTM. Still open review comments can be addressed later, it's nice this P-R moves code but does not change the implementation.

maass-hamburg commented:
> @maass-hamburg Ping on this. Also, should I consider it as a blocking comment ?

No, not blocking, I'm not really in favor, but I can live with it

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v1.c:
```diff
@@ -2,88 +2,29 @@
-#define __eth_stm32_desc __aligned(4)
-#define __eth_stm32_buf  __aligned(4)
-#endif
+LOG_MODULE_DECLARE(eth_stm32_hal);
```
> Why?

My mistake. Discard my question :confused:.
I think @erwango addressed your review comment.

erwango commented:
@maass-hamburg Non regression campaign was passed successfully. PTAL.

pdgendt commented:
> @pdgendt If you're interested.

We can test these changes somewhere in the coming days, should not block the PR though. Will report any issues if encountered.
